nomad/state/state_store.go

```diff
@@ -437,6 +449,86 @@ func (s *StateStore) DeleteJob(index uint64, jobID string) error {
 	return nil
 }
 
+// deleteJobVersions deletes all versions of the given job.
+func (s *StateStore) deleteJobVersions(index uint64, job *structs.Job, txn *memdb.Txn) error {
+	iter, err := txn.Get("job_versions", "id_prefix", job.ID)
+	if err != nil {
+		return err
+	}
+
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+
+		// Ensure the ID is an exact match
+		j := raw.(*structs.Job)
+		if j.ID != job.ID {
+			continue
+		}
+
+		if _, err = txn.DeleteAll("job_versions", "id", job.ID, job.Version); err != nil {
```

Comment:
Because we are doing a prefix match to get all the versions for the ID you may actually delete another job that starts with the same prefix.